gbpservice/nfp/service_plugins/vpn/drivers/nfp_vpnaaas_driver.py

```diff
@@ -0,0 +1,418 @@
+        the agent which is hosting this vpn service
+        """
+        host = agent['host']
+        lhost = socket.gethostname()
+        if host == lhost:
+            return True
+        return False
+
+    def _get_agent_hosting_vpnservice(self, admin_context, vpnservice_id):
+        filters = {'agent_type': [const.AGENT_TYPE_VPN]}
+        agents = manager.NeutronManager.get_plugin().get_agents(
+            admin_context,  filters=filters)
+
+        try:
+            for agent in agents:
+                if not agent['alive']:
+                    continue
+                res = self._is_agent_hosting_vpnservice(
+                    agent, vpnservice_id)
+                if res is True:
+                    return agent
+
+            # valid vpn agent is not found, hostname comparison might be
+            # failed. Return whichever agent is available.
+            for agent in agents:
+                if not agent['alive']:
+                    continue
+                return agent
+        except:
+            raise VPNAgentNotFound()
+
+        LOG.error(_('No active vpn agent found. Configuration will fail.'))
+        raise VPNAgentHostingServiceNotFound(vpnservice_id=vpnservice_id)
+
+    def _agent_notification(self, context, method, vpnservice_id,
+                            version=None, **kwargs):
+        """Notify update for the agent.
+            For some reason search with
+            'agent_type=AGENT_TYPE_VPN is not working.
+            Hence, get all the agents,
+            loop and find AGENT_TYPE_VPN, and also the one which
+            is hosting that vpn service and
+            implementing OC_IPSEC_TOPIC
+        """
+        admin_context = context.is_admin and context or context.elevated()
+
+        if not version:
+            version = self.target.version
+        vpn_agent = self._get_agent_hosting_vpnservice(
+            admin_context, vpnservice_id)
+
+        LOG.debug(_('Notify agent at %(topic)s.%(host)s the message '
+                    '%(method)s %(args)s'), {
+            'topic': self.topic, 'host': vpn_agent['host'],
+            'method': method, 'args': kwargs})
+
+        cctxt = self.client.prepare(server=vpn_agent['host'],
+                                    version=version)
+        cctxt.cast(context, method, **kwargs)
+
+    def vpnservice_updated(self, context, vpnservice_id, **kwargs):
+        """
+        Make rpc to agent for 'vpnservice_updated'
+        """
+        try:
+            self._agent_notification(
+                context, 'vpnservice_updated',
+                vpnservice_id, **kwargs)
+        except:
+            LOG.error(_('Notifying agent failed'))
+
+
+class OCIPsecVPNDriver(base_ipsec.BaseIPsecVPNDriver):
+    """VPN Service Driver class for IPsec."""
+
+    def __init__(self, service_plugin):
+        super(OCIPsecVPNDriver, self).__init__(
+            service_plugin,
+            OCVpnValidator(service_plugin))
```

Comment:
Use validator=None
Right now we don't have any validator. The class OCVpnValidator is unused.